util/app/discovery/discovery.go

```diff
@@ -99,109 +90,167 @@ func DetectConfigManagementPlugin(ctx context.Context, appPath, repoPath, plugin
+	cmpClient, closer, isDiscoveryConfigured, err := getClientForFilepath(ctx, c)
 	if err != nil {
-		return false, false, fmt.Errorf("error getting stream client: %w", err)
+		log.Errorf("error getting client for plugin: %v", err)
+		return nil, nil
 	}
+	if !isDiscoveryConfigured {
+		// If discovery isn't configured, assume the CMP supports the plugin since it was explicitly named.
+		return cmpClient, closer
+	}
+	usePlugin := cmpSupportsForClient(ctx, cmpClient, appPath, repoPath, env, tarExcludedGlobs)
+	if !usePlugin {
+		io.Close(closer)
+		return nil, nil
+	}
+	return cmpClient, closer
+}
 
-	err = cmp.SendRepoStream(ctx, appPath, repoPath, matchRepoStream, env, tarExcludedGlobs)
+// unnamedCMPSupports will return a client if the plugin supports the given repository. It will also return a closer for
+// the connection.
+func unnamedCMPSupports(ctx context.Context, c CMPClientConstructor, appPath, repoPath string, env, tarExcludedGlobs []string) (pluginclient.ConfigManagementPluginServiceClient, io.Closer) {
+	cmpClient, closer, isDiscoveryConfigured, err := getClientForFilepath(ctx, c)
 	if err != nil {
-		return false, false, fmt.Errorf("error sending stream: %w", err)
+		log.Errorf("error getting client for plugin: %v", err)
+		return nil, nil
 	}
-	resp, err := matchRepoStream.CloseAndRecv()
-	if err != nil {
-		return false, false, fmt.Errorf("error receiving stream response: %w", err)
+	if !isDiscoveryConfigured {
+		// If discovery isn't configured, we can't assume the CMP supports the repo.
+		return nil, nil
+	}
+	usePlugin := cmpSupportsForClient(ctx, cmpClient, appPath, repoPath, env, tarExcludedGlobs)
+	if !usePlugin {
+		io.Close(closer)
+		return nil, nil
 	}
-	return resp.GetIsSupported(), resp.GetIsDiscoveryEnabled(), nil
+	return cmpClient, closer
 }
 
-func cmpSupports(ctx context.Context, pluginSockFilePath, appPath, repoPath, fileName string, env []string, tarExcludedGlobs []string, namedPlugin bool) (io.Closer, pluginclient.ConfigManagementPluginServiceClient, bool) {
-	absPluginSockFilePath, err := filepath.Abs(pluginSockFilePath)
+// CMPClientConstructor is an interface for creating a client for a CMP.
+type CMPClientConstructor interface {
+	// NewConfigManagementPluginClient returns a client for the CMP. It also returns a closer for the connection.
+	NewConfigManagementPluginClient() (pluginclient.ConfigManagementPluginServiceClient, io.Closer, error)
+}
+
+// socketCMPClientConstructor is a CMPClientConstructor for a CMP that uses a socket file.
+type socketCMPClientConstructor struct {
+	pluginSockFilePath string
+	filename           string
+}
+
+// newSocketCMPClientConstructorForPath returns a new socketCMPClientConstructor.
+func newSocketCMPClientConstructorForPath(pluginSockFilePath, filename string) CMPClientConstructor {
```

Comment:
Just an OOP tip but as you constructing a `socketCMPClientConstructor` most of the times it preferable to return the concrete type than the interface. 2 main reasons are:
- The type can have additional methods that wouldn't be available if we return the interface
- The function says `newSocketCMPClientConstructorForPath` but returns an interface instead.

This is a nice write up about it:
https://dev.to/lcaparelli/should-my-methods-return-structs-or-interfaces-in-go-3b7

Reply:
Good tip. Fixed!